src/libos/crates/async-mountfs/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+        let weak = Arc::downgrade(&fs);
+        let ptr = Arc::into_raw(fs) as *mut Self;
+        unsafe {
+            (*ptr).self_ref = weak;
+            Arc::from_raw(ptr)
+        }
+    }
+
+    async fn mountpoint_root_inode(&self) -> Arc<AsyncMInode> {
+        AsyncMInode::new(
+            self.inner.root_inode().await,
+            self.self_ref.upgrade().unwrap(),
+        )
+    }
+}
+
+#[async_trait]
+impl AsyncFileSystem for AsyncMountFS {
+    async fn sync(&self) -> Result<()> {
+        self.inner.sync().await?;
+        for mount_fs in self.mountpoints.read().await.values() {
+            mount_fs.sync().await?;
+        }
+        Ok(())
+    }
+
+    async fn root_inode(&self) -> Arc<dyn AsyncInode> {
+        match &self.self_mountpoint {
+            Some(inode) => inode.vfs.root_inode().await,
+            None => self.mountpoint_root_inode().await,
+        }
+    }
+
+    async fn info(&self) -> FsInfo {
+        self.inner.info().await
+    }
+}
+
+/// Inode for `AsyncMountFS`
+pub struct AsyncMInode {
+    /// The inner Inode
+    inner: Arc<dyn AsyncInode>,
+    /// Associated `AsyncMountFS`
+    vfs: Arc<AsyncMountFS>,
+    /// Weak reference to self
+    self_ref: Weak<AsyncMInode>,
+}
+
+impl AsyncMInode {
+    pub fn new(inode: Arc<dyn AsyncInode>, vfs: Arc<AsyncMountFS>) -> Arc<Self> {
+        Self {
+            inner: inode,
+            vfs,
+            self_ref: Weak::default(),
+        }
+        .wrap()
+    }
+
+    fn wrap(self) -> Arc<Self> {
+        let inode = Arc::new(self);
+        let weak = Arc::downgrade(&inode);
+        let ptr = Arc::into_raw(inode) as *mut Self;
+        unsafe {
+            (*ptr).self_ref = weak;
+            Arc::from_raw(ptr)
+        }
+    }
+
+    /// Get the root inode of the mounted fs at here.
+    /// Return self if no mounted fs.
+    async fn overlaid_inode(&self) -> Arc<AsyncMInode> {
+        let inode_id = self.metadata().await.unwrap().inode;
+        if let Some(sub_vfs) = self.vfs.mountpoints.read().await.get(&inode_id) {
+            sub_vfs.mountpoint_root_inode().await
+        } else {
+            self.self_ref.upgrade().unwrap()
+        }
+    }
+
+    /// If is the root inode of its fs
+    async fn is_mountpoint_root(&self) -> bool {
+        self.inner
+            .fs()
+            .root_inode()
+            .await
+            .metadata()
+            .await
+            .unwrap()
+            .inode
+            == self.inner.metadata().await.unwrap().inode
+    }
+
+    /// Strong type version of `create()`
+    pub async fn create(&self, name: &str, type_: FileType, mode: u16) -> Result<Arc<Self>> {
+        let inode = Self::new(
+            self.inner.create(name, type_, mode).await?,
+            self.vfs.clone(),
+        );
+        Ok(inode)
+    }
+
+    /// Strong type version of `find()`
+    pub async fn find(&self, name: &str) -> Result<Arc<Self>> {
+        let mut result = self.self_ref.upgrade().unwrap();
```

Comment:
Choose more descriptive names, e.g., `arc_self` and `this`.

Reply:
Done. Thanks